### PR TITLE
S2A record protocol refactoring to use key derivation

### PR DESCRIPTION
**Pull request by matthewstevenson88**
This PR refactors the `s2a_crypter_create` method to derive the keys and nonces from the traffic secret. It depends on #4 and #5, but it is orthogonal to the changes made in #5.

Note: because the branch s2a-encrypt is slightly out of sync with the branch s2a-rp-create, this PR may show changes to files in the directory src/core/tsi/alts/crypt. Please ignore them. They will disappear when #4 is merged with s2a-master.

### CODE REVIEW

File: test/core/tsi/s2a/s2a_test_util.cc [hunk 1]
```diff
@@ -105,6 +106,36 @@ std::vector<uint8_t> chacha_poly_empty_record_bytes = {
...
+      break;
+    default:
+      gpr_log(GPR_ERROR, kS2AUnsupportedCiphersuite);
+      abort();
+  }
+  check_half_connection(
+      crypter, /** in_half_connection **/ true,
+      /** expected_sequence **/ 0, expected_traffic_secret.size(),
+      expected_traffic_secret.data(),
+      /** expected nonce size **/ 12, expected_nonce, SSL3_RT_HEADER_LENGTH);
```
Comment (cesarghali) [inline test/core/tsi/s2a/s2a_test_util.cc]: Do we have a const for `12`? It can be local to this file or function.

File: test/core/tsi/s2a/BUILD [hunk 2]
```diff
@@ -21,7 +21,9 @@ grpc_package(name = "test/core/tsi/s2a")
 grpc_cc_library(
     name = "s2a_test_util",
     srcs = ["s2a_test_util.cc"],
-    hdrs = ["s2a_test_util.h"],
+    hdrs = [
```
Comment (cesarghali) [inline test/core/tsi/s2a/BUILD]: Is this auto-generated? If not please revert this change to keep the PR simple, unless it was inherited from an upstream merge.

File: src/core/tsi/s2a/record_protocol/s2a_crypter.cc [hunk 3]
```diff
@@ -86,27 +88,135 @@ static grpc_status_code s2a_tag_size(const s2a_crypter* crypter,
...
+ *  - output_size: the number of bytes to be written to the |output| buffer.
+ *  - output: the buffer where the derived secret is written; this is owned by
+ *    the caller.
+ *  - error_details: an error message for when the creation fails. It is legal
+ *    (and expected) to have |error_details| point to a nullptr; otherwise,
+ *    the argument should be freed with gpr_free.
+ *
+ *  On success, the method returns GRPC_STATUS_OK. Otherwise, it returns an
+ * error status code and details can be found in |error_details|. **/
+static grpc_status_code derive_tls13_secret(uint16_t ciphersuite, bool is_key,
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/s2a/record_protocol/s2a_crypter.cc]: Done in next commit.

File: src/core/tsi/s2a/record_protocol/s2a_crypter.cc [hunk 4]
```diff
@@ -86,27 +88,135 @@ static grpc_status_code s2a_tag_size(const s2a_crypter* crypter,
...
+      break;
+    case kTlsChacha20Poly1305Sha256:
+      hash_function = GsecHashFunction::SHA256_hash_function;
+      break;
+    default:
+      *error_details = gpr_strdup(kS2AUnsupportedCiphersuite);
+      return GRPC_STATUS_FAILED_PRECONDITION;
+  }
+
+  uint8_t label[2 + 11];
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/s2a/record_protocol/s2a_crypter.cc]: Done in next comment.

File: test/core/tsi/s2a/BUILD (hunk 2 above)
Comment (matthewstevenson88) [reply test/core/tsi/s2a/BUILD]: This was auto-generated. Not sure why it makes this change sometimes and leaves it others.

File: test/core/tsi/s2a/s2a_test_util.cc (hunk 1 above)
Comment (matthewstevenson88) [reply test/core/tsi/s2a/s2a_test_util.cc]: In the next commit, I've deduced the expected nonce size from the test vector (rather than using `12` directly).

**Comment by matthewstevenson88** [comment]
Thanks for the review @cesarghali! I've implemented the changes you requested.

FYI I also removed the null byte from the test vectors that started their lives as strings.

**Comment by matthewstevenson88** [comment]
Thanks for the quick review @cesarghali! :)
